### PR TITLE
feat: add resume option

### LINKED ISSUE #225: Add a resume mode for tdp-lib
We could use to have a `--resume` mode in `tdp deploy`.

Without parameter, it would resume the last deploy action using the same DAG, skipping the successful nodes. With a parameter it would resume a given deployment id.

An attempt to resume a Successful deployment should raise an exception.

### PR DESCRIPTION
**Pull request by leopaul36**
#### Which issue(s) this PR fixes:

Fixes #225

Main features:
- Add `--resume` option
- Move `get_deployment` to `queries.py`
- Introduced new Enum `DeploymentTypeEnum`
- Also applied the non resolved comments from https://github.com/TOSIT-IO/tdp-lib/pull/244

#### Additional comments:

To make clear that you license your contribution under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0) and that you give permission to [TOSIT](https://www.tosit.io/), you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licensed under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0).
 - [x] I hereby agree to grant [TOSIT](https://www.tosit.io/) a copyright license to use my contributions.

### CODE REVIEW

File: tdp/cli/commands/deploy.py [hunk 1]
```diff
@@ -62,6 +64,12 @@
 )
 @click.option("--filter", type=str, help="Glob on list name")
 @click.option("--dry", is_flag=True, help="Execute dag without running any operation")
+@click.option(
+    "--resume",
+    type=str,
+    metavar="dep_id",
+    help="Resume a previous deployment, cannot be used with --sources or --targets",
```
Comment (rpignolet) [inline tdp/cli/commands/deploy.py:71]: `--filter` is allowed ? If the "resumed deployment" had a `filter`, the `--filter` will be overridden. I think `--filter` cannot be used with `resume` too.

File: tdp/cli/commands/deploy.py [hunk 2]
```diff
@@ -99,16 +108,62 @@ def deploy(
...
+                ]
+                # print(original_operations)
+                succeeded_operations = [
+                    operation.operation
+                    for operation in resumed_deployment.operations
+                    if operation.state == StateEnum.SUCCESS
+                ]
+                # print(succeeded_operations)
+                remaining_operations_list = list(
+                    set(original_operations) - set(succeeded_operations)
```
Comment (rpignolet) [inline tdp/cli/commands/deploy.py:154]: This logic can not be moved in `DeploymentPlan` class ? We can add a method in `DeploymentPlan` which take a list of operations and change the plan with only remaining operations. Or maybe the method can take a `DeploymentLog` instance.

**Review by rpignolet — CHANGES_REQUESTED** [review]

File: tdp/cli/commands/deploy.py [hunk 3]
```diff
@@ -62,6 +64,12 @@
 )
 @click.option("--filter", type=str, help="Glob on list name")
 @click.option("--dry", is_flag=True, help="Execute dag without running any operation")
+@click.option(
```
Comment (gboutry) [inline tdp/cli/commands/deploy.py:67]: Does this mean we need to pass the deployment id anytime we want to resume ? Can't we resume automatically if previous deployment was a failure ?

Comment (leopaul36) [reply tdp/cli/commands/deploy.py:67]: We can do this, would a `get_last_deployment` with a query like `select(DeploymentLog).order_by(DeploymentLog.id.desc()).first()` be fine?

Comment (gboutry) [reply tdp/cli/commands/deploy.py:67]: Yes

File: tdp/core/runner/deployment_plan.py [hunk 4]
```diff
@@ -51,6 +52,14 @@ def from_dag(
     def from_operations(operations):
         deployment_args = dict(
             targets=[operation.name for operation in operations],
-            using_dag=False,
+            deployment_type=DeploymentTypeEnum.OPERATIONS,
+        )
+        return DeploymentPlan(operations, deployment_args)
+
+    @staticmethod
```
Comment (gboutry) [inline tdp/core/runner/deployment_plan.py:59]: This should take a DeploymentLog as input (not operations), and should fail if the deploymentlog is in success

**Comment by leopaul36** [comment]
Closed in favor of https://github.com/TOSIT-IO/tdp-lib/pull/258

**Pull request closed without merging**